File: nodejs/aws/logger/lib/cloud-watch-logger.js
const AWS = require("aws-sdk");

const { Exception, Logger } = require("@mcma/core");

const CloudWatchLogs = new AWS.CloudWatchLogs();

class AwsCloudWatchLogger extends Logger {
    constructor(source, tracker, addLogEvent) {
        super(source, tracker);

        this.log = (level, type, msg, ...args) => {
            if (level > 0) {
                const timestamp = Date.now();
                const message = this.buildLogEvent(timestamp, level, type, msg, ...args);

                addLogEvent(message, timestamp);
            }
        };
    }
}

class AwsCloudWatchLoggerProvider {
    constructor(source, logGroupName) {
        if (typeof source !== "string" || typeof logGroupName !== "string") {
            throw new Exception("Failed to initialize AwsCloudWatchLoggerProvider with params source: '" + source + "' and logGroupName: '" + logGroupName + "'");
        }

        const logStreamName = source + "-" + Date.now();

        let logEvents = [];
        let logGroupVerified = false;
        let logStreamCreated = false;
        let processing = false;
        let sequenceToken = undefined;

        const processBatch = async () => {
            if (processing) {
                return;
            }
            try {
                processing = true;

                while (logEvents.length) {
                    if (!logGroupVerified) {
                        let nextToken = undefined;

                        do {
                            const params = {
                                logGroupNamePrefix: logGroupName,
                                nextToken,
                            };

                            const data = await CloudWatchLogs.describeLogGroups(params).promise();

                            for (const logGroup of data.logGroups) {
                                if (logGroup.logGroupName === logGroupName) {
                                    logGroupVerified = true;
                                    break;
                                }
                            }

                            nextToken = data.nextToken;
                        } while (!logGroupVerified && nextToken);

                        if (!logGroupVerified) {

                            const params = {
                                logGroupName
                            };

                            await CloudWatchLogs.createLogGroup(params).promise();
                            logGroupVerified = true;
                        }
                    }

                    if (!logStreamCreated) {
                        const params = {
                            logGroupName,
                            logStreamName
                        };

                        await CloudWatchLogs.createLogStream(params).promise();

                        logStreamCreated = true;
                    }

                    const params = {
                        logEvents,
                        logGroupName,
                        logStreamName,
                        sequenceToken
                    };

                    logEvents = [];

                    const data = await CloudWatchLogs.putLogEvents(params).promise();

                    sequenceToken = data.nextSequenceToken;

                    if (data.rejectedLogEventsInfo) {
                        console.error("AwsCloudWatchLogger: Some log events rejected");
                        console.error(data.rejectedLogEventsInfo);
                    }
                }
            } catch (error) {
                console.error("AwsCloudWatchLogger: Failed to log to CloudWatchLogs");
                console.error(error);
            } finally {
                processing = false;
            }
        };

        const addLogEvent = (message, timestamp) => {
            logEvents.push({ message, timestamp });
            processBatch();
        };

        this.get = (tracker) => new AwsCloudWatchLogger(source, tracker, addLogEvent);
    }
}

module.exports = {
    AwsCloudWatchLoggerProvider
};

File: dotnet/aws/Mcma.Aws.CloudWatch/CloudWatchLogger.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Amazon.CloudWatchLogs;
using Mcma.Core.Logging;

using McmaLogger = Mcma.Core.Logging.Logger;
using IMcmaLogger = Mcma.Core.Logging.ILogger;
using Amazon.CloudWatchLogs.Model;
using System.Linq;
using Mcma.Core;
using Mcma.Core.Serialization;

namespace Mcma.Aws.CloudWatch
{
    public class AwsCloudWatchLogger : McmaLogger
    {
        public AwsCloudWatchLogger(string source, McmaTracker tracker, Action<LogEvent> log)
            : base(source, tracker)
        {

        }

        protected override void Log(LogEvent logEvent)
        {
        }
    }

    public class AwsCloudWatchLoggerProvider : LoggerProvider<AwsCloudWatchLogger>
    {
        public AwsCloudWatchLoggerProvider(string source, string logGroupName)
            : base(source)
        {
            LogGroupName = logGroupName;
            LogStreamName = source + "-" + Guid.NewGuid();
        }

        private AmazonCloudWatchLogsClient CloudWatchLogsClient { get; } = new AmazonCloudWatchLogsClient();

        private string LogGroupName { get; }

        private string LogStreamName { get; }

        private List<InputLogEvent> LogEvents { get; } = new List<InputLogEvent>();

        private object LogEventsLock { get; } = new object();

        private bool LogGroupCreated { get; set; }
        
        private bool LogStreamCreated { get; set; }

        private Task ProcessingTask { get; set; }
        
        private string SequenceToken { get; set; }

        private async Task ProcessBatchAsync()
        {
            try
            {
                await EnsureLogGroupAndStreamCreatedAsync();
                
                List<InputLogEvent> logEvents;
                lock (LogEventsLock)
                {
                    logEvents = LogEvents.ToList();
                    LogEvents.Clear();
                }

                while (logEvents.Count > 0)
                {
                    var request = new PutLogEventsRequest
                    {
                        LogEvents = logEvents,
                        LogGroupName = LogGroupName,
                        LogStreamName = LogStreamName,
                        SequenceToken = SequenceToken
                    };

                    var data = await CloudWatchLogsClient.PutLogEventsAsync(request);

                    SequenceToken = data.NextSequenceToken;

                    if (data.RejectedLogEventsInfo != null)
                        Logger.System.Error("AwsCloudWatchLogger: Some log events rejected", data.RejectedLogEventsInfo);
                }
            }
            catch (Exception error)
            {
                Logger.System.Error("AwsCloudWatchLogger: Failed to log to CloudWatchLogs", error);
            }
        }

        private async Task EnsureLogGroupAndStreamCreatedAsync()
        {
            if (LogGroupCreated && LogStreamCreated)
                return;

            if (!LogGroupCreated)
            {
                string nextToken = null;

                do
                {
                    var request = new DescribeLogGroupsRequest
                    {
                        LogGroupNamePrefix = LogGroupName,
                        NextToken = nextToken
                    };

                    var data = await CloudWatchLogsClient.DescribeLogGroupsAsync(request);

                    foreach (var logGroup in data.LogGroups)
                    {
                        if (logGroup.LogGroupName == LogGroupName)
                        {
                            LogGroupCreated = true;
                            break;
                        }
                    }

                    nextToken = data.NextToken;
                } while (!LogGroupCreated && !string.IsNullOrWhiteSpace(nextToken));

                if (!LogGroupCreated) {

                    var request = new CreateLogGroupRequest { LogGroupName = LogGroupName };

                    await CloudWatchLogsClient.CreateLogGroupAsync(request);
                    LogGroupCreated = true;
                }
            }
            
            if (!LogStreamCreated)
            {
                var request = new CreateLogStreamRequest
                {
                    LogGroupName = LogGroupName,
                    LogStreamName = LogStreamName
                };

                await CloudWatchLogsClient.CreateLogStreamAsync(request);

                LogStreamCreated = true;
            }
        }

        private void AddLogEvent(LogEvent logEvent)
        {
            lock (LogEventsLock)
                LogEvents.Add(new InputLogEvent { Message = logEvent.ToMcmaJson().ToString(), Timestamp = logEvent.Timestamp.DateTime });

            if (ProcessingTask == null)
            {
                ProcessingTask = ProcessBatchAsync();
                ProcessingTask.ContinueWith(t => ProcessingTask = null);
            }
        }

        protected override AwsCloudWatchLogger Get(string source, McmaTracker tracker = null) => new AwsCloudWatchLogger(source, tracker, AddLogEvent);

        public async Task FlushAsync()
        {
            if (ProcessingTask != null)
                await ProcessingTask;
        }
    }
}